File: src/utils/inv_stack_batch.py
#!/usr/bin/python
import xml.etree.ElementTree as ET
import os
from datetime import datetime
from socket import gethostname

class BatchInfo(object):

    def __init__(self, settings_dir, settings_tags, out_parent_dir):
        self.timestamp = datetime.today().isoformat('_').split('.')[0]
        self.settings_dir = settings_dir
        self.out_dir = os.path.join(out_parent_dir, self.timestamp)
        self.settings_dict = {}
        self.env_dict = {}

        self.make_out_dir()
        self.set_settings_dict(settings_tags)
        self.set_env_dict()

    def make_out_dir(self):
        if not os.path.exists(self.out_dir):
            os.makedirs(self.out_dir)

    def set_settings_dict(self, settings_tags):
        tree = ET.parse(os.path.join(self.settings_dir,'settings.xml'))
        for tag in settings_tags:
            element = tree.find(tag)
            self.settings_dict[element.tag] = element.text

    def set_env_dict(self):
        self.env_dict['hostname'] = gethostname()

    def write(self, out_file='batch_info.txt'):
        fmt = '%s\t%s\n'
        fh = open(os.path.join(self.out_dir, out_file), 'w')
        fh.write(fmt % ('timestamp', self.timestamp))
        for k in sorted(self.env_dict.keys()):
            fh.write(fmt % (k, self.env_dict[k]))
        for k in sorted(self.settings_dict.keys()):
            fh.write(fmt % (k, self.settings_dict[k]))
        fh.close

#class RunInfo

batch = BatchInfo(settings_dir = r'/home/rahaman/benchmarks/mc-performance/openmc', 
        out_parent_dir = './profiles', 
        settings_tags = ( 'eigenvalue/batches', 'eigenvalue/inactive',
            'eigenvalue/particles', 'energy_grid'))

batch.write()

# # Values for interpolation thershholds
# thresh_vals = (0, 0.0125, 0.025, 0.05, 0.1, 0.15, 0.2, 0.3, 0.4, 0.6, 0.8, 0.875, 0.9, 0.9125)

# # Locatiion for nuclides_info file
# nuc_info_dir = r'data_dump/'
# nuc_info_file = r'nuclides_info.txt'

File: src/inv_stack_batch.pl
#!/usr/bin/perl
use strict;

$, = "\t";
$\ = "\n";

# Values for interpolation thershholds
my @thresh_vals = qw(0 0.0125 0.025 0.05 0.1 0.15 0.2 0.3 0.4 0.6 0.8 0.875 0.9 0.9125);

# Locations for settings.xml
my $settings_dir = '/home/rahaman/benchmarks/mc-performance/openmc/';
my $settings_file = 'settings.xml';

# Locatiion for nuclides_info file
my $nuc_info_dir = 'data_dump/';
my $nuc_info_file = 'nuclides_info.txt';

# Location for profiles
my $gprof_dir = 'profiles/';

# Open batch results file and print header
open my $result_fh, '>', $gprof_dir.'batch_result.txt'
  or die "Couldn't open batch_result.txt ; $!";
print $result_fh qw(n_particles interp_thresh count_ngrid mb_ngrid k_eff
  n_collisions n_col_per_part calc_rate);

# Get number of particles from settings file
my $n_particles = sum_particles($settings_dir.$settings_file);

# For each interp_thresh, run OpenMC and parse output
for my $interp_thresh (@thresh_vals) {
  my ($calc_rate, $k_eff, $count_ngrid, $mb_ngrid, $n_collisions);

  # run OpenMC and capture calculation rate, k-effective from STDOUT
  print "*** BEGIN OPENMC WITH INTERP_THRESH = $interp_thresh";
  open my $openmc_fh, '-|',  "./openmc $settings_dir $interp_thresh"
    or die "Couldn't open filehandle for openmc; $!";
  while (<$openmc_fh>) {
    if (/Calculation Rate \(active\)\s+=\s+(\d+\.\d+)/) {
      $calc_rate = $1;
    }
    elsif (/k-effective \(Track-length\)\s+=\s+(\d+\.\d+)/) {
      $k_eff = $1;
    }
  }

  # Parse sum(n_grid) from data dump
  $count_ngrid = count_ngrid($nuc_info_dir.$nuc_info_file);
  $mb_ngrid = $count_ngrid*8/1048576;

  # run gprof
  my $gprof_file = sprintf("grid_%dMB_%dpts.profile", $mb_ngrid, $count_ngrid);
  #my $gprof_file = "npart_$n_particles.ngrid_$count_ngrid.profile";
  system("gprof ./openmc > ".$gprof_dir.$gprof_file);

  # Parse n_collisons from gprof output
  $n_collisions = n_collisions($gprof_dir.$gprof_file);

  print $result_fh ($n_particles, $interp_thresh, $count_ngrid, $mb_ngrid,
    $k_eff, $n_collisions, $n_collisions/$n_particles, $calc_rate);
}

close $result_fh;

sub sum_particles {
  my ($n_particles, $n_batches);
  # Open the first argument as it was a filename
  print $_[0];
  open my $fh, '<', $_[0] or die "Error in sum_particles: $!";
  while (<$fh>) {
    if (/particles\>(\d+)\<\/particles/) {
      $n_particles = $1;
    }
    elsif (/batches\>(\d+)\<\/batches/) {
      $n_batches = $1;
    }
  }
  close $fh;
  return $n_particles * $n_batches;
}

sub count_ngrid {
  # Open the first argument as it was a filename
  open my $fh, '<', $_[0] or die "Error in count_ngrid: $!";
  # Sum the second column in the file
  my $sum = 0;
  while(<$fh>) {
    chomp;
    my @linelist = split;
    $sum = $sum + $linelist[1]*6;
  }
  close $fh;
  return $sum;
}

sub n_collisions {
  # Open the first argument as it was a filename
  open my $fh, '<', $_[0] or die "Error in n_collisions: $!";
  while (<$fh>) {
    if (/collision$/) {
      my @linelist = split;
      close $fh;
      return $linelist[3];
    }
  }
  close $fh;
  return -1;
}

